File: src/backstop/audit.py
"""Tamper-evident audit log (Deep Research P2#8).

Every enforcement decision (deny / fallback / downgrade / cache_hit) is appended
as a JSON record whose integrity is chained: each record carries an HMAC of its
own payload plus the previous record's chain hash. `verify()` replays the chain
and detects any tampering. This is the enterprise "escape hatch" that makes
in-process enforcement audit-ready and closes the supply-chain / audit gap the
research flagged (e.g. the March-2026 LiteLLM incident class).
"""
from __future__ import annotations

import hashlib
import hmac
import json
import threading
import time
from typing import Any, Callable


def _chain_hash(prev: bytes, payload: bytes, key: bytes) -> str:
    mac = hmac.new(key, prev + payload, hashlib.sha256).hexdigest()
    return hashlib.sha256((prev.hex() + mac).encode()).hexdigest()


class AuditLog:
    def __init__(self, sink: str | Callable[[str], None], hmac_key: str | None = None) -> None:
        self._key = (hmac_key or "").encode("utf-8")
        self._prev = b""
        self._lock = threading.Lock()
        self._file = open(sink, "a", encoding="utf-8") if isinstance(sink, str) else None
        self._callable = sink if callable(sink) else None

    def record(self, decision: str, reason: str, **fields: Any) -> dict:
        rec = {"ts": time.time(), "decision": decision, "reason": reason}
        if fields:
            rec.update(fields)
        payload = json.dumps(rec, sort_keys=True, default=str).encode("utf-8")
        with self._lock:
            chain = _chain_hash(self._prev, payload, self._key)
            rec["_chain"] = chain
            self._prev = bytes.fromhex(chain)
        line = json.dumps(rec, sort_keys=True, default=str)
        with self._lock:
            if self._file is not None:
                self._file.write(line + "\n")
                self._file.flush()
            elif self._callable is not None:
                self._callable(line)
        return rec

    def verify(self, lines: list[str] | None = None) -> bool:
        prev = b""
        source = lines if lines is not None else self._read_lines()
        for line in source:
            line = line.strip()
            if not line:
                continue
            rec = json.loads(line)
            payload = json.dumps(
                {k: v for k, v in rec.items() if k != "_chain"}, sort_keys=True, default=str
            ).encode("utf-8")
            chain = _chain_hash(prev, payload, self._key)
            if chain != rec.get("_chain"):
                return False
            prev = bytes.fromhex(rec["_chain"])
        return True

    def _read_lines(self) -> list[str]:
        if self._file is None:
            return []
        path = self._file.name
        self._file.flush()
        with open(path, "r", encoding="utf-8") as fh:
            return fh.readlines()

    def close(self) -> None:
        if self._file is not None:
            self._file.close()
            self._file = None

File: ts/backstop/src/audit.ts
import { createHash, createHmac } from "node:crypto";
import type { AuditSink } from "./types.js";

export interface AuditRecord {
  ts: number;
  decision: string;
  reason: string;
  _chain: string;
  [key: string]: unknown;
}

function chainHash(prev: Buffer, payload: Buffer, key: Buffer): string {
  const mac = createHmac("sha256", key).update(Buffer.concat([prev, payload])).digest();
  return createHash("sha256").update(Buffer.concat([prev, mac])).digest("hex");
}

/**
 * Tamper-evident audit log. Mirrors the Python AuditLog.
 */
export class AuditLog {
  private key: Buffer;
  private prev = Buffer.alloc(0);
  private file: NodeJS.WritableStream | null = null;
  private callable: ((line: string) => void) | null = null;
  private lines: string[] = [];

  constructor(sink: AuditSink | null, hmacKey?: string) {
    this.key = Buffer.from(hmacKey ?? "", "utf-8");
    if (typeof sink === "string") {
      const fs = require("node:fs") as typeof import("node:fs");
      this.file = fs.createWriteStream(sink, { flags: "a" });
    } else if (typeof sink === "function") {
      this.callable = sink;
    }
  }

  record(decision: string, reason: string, fields: Record<string, unknown> = {}): AuditRecord {
    const rec = { ts: Date.now(), decision, reason, ...fields };
    const payload = Buffer.from(JSON.stringify(rec), "utf-8");
    const chain = chainHash(this.prev, payload, this.key);
    const full = { ...rec, _chain: chain };
    this.prev = Buffer.from(chain, "hex");
    const line = JSON.stringify(full);
    this.lines.push(line);
    if (this.file) this.file.write(line + "\n");
    if (this.callable) this.callable(line);
    return full;
  }

  verify(lines?: string[]): boolean {
    let prev = Buffer.alloc(0);
    const src = lines ?? [...this.lines];
    for (const line of src) {
      const rec = JSON.parse(line);
      const { _chain, ...rest } = rec;
      const payload = Buffer.from(JSON.stringify(rest), "utf-8");
      const expected = chainHash(prev, payload, this.key);
      if (expected !== _chain) return false;
      prev = Buffer.from(_chain, "hex");
    }
    return true;
  }

  close(): void {
    if (this.file) {
      this.file.end();
      this.file = null;
    }
  }
}
